projects/storefrontlib/src/cms-components/myaccount/update-email/update-email.service.ts

```diff
@@ -0,0 +1,94 @@
+import { Injectable, OnDestroy } from '@angular/core';
+import {
+  AuthService,
+  GlobalMessageService,
+  GlobalMessageType,
+  RoutingService,
+  UserService,
+} from '@spartacus/core';
+import { Observable, Subscription } from 'rxjs';
+import { FormControl, FormGroup, Validators } from '@angular/forms';
+import { CustomFormValidators } from '../../../shared/utils';
+
+@Injectable({
+  providedIn: 'root'
+})
+
+export class UpdateEmailService implements OnDestroy {
+  protected subscriptions = new Subscription();
+  protected newUid: string;
+
+  constructor(
+    protected routingService: RoutingService,
+    protected globalMessageService: GlobalMessageService,
+    protected userService: UserService,
+    protected authService: AuthService
+  ) {
+  }
+
+  form: FormGroup = new FormGroup(
+    {
+      email: new FormControl('', [Validators.required, CustomFormValidators.emailValidator]),
+      confirmEmail: new FormControl('', [Validators.required]),
+      password: new FormControl('', [Validators.required])
+    },
+    {
+    validators: CustomFormValidators.emailsMustMatch('email', 'confirmEmail')
+  });
+
+  getUpdateEmailResultLoading(): Observable<boolean> {
+    return this.userService.getUpdateEmailResultLoading();
+  }
+
+  resetUpdateEmailResultState(): void {
+    this.userService.resetUpdateEmailResultState();
+  }
+
+  ngOnDestroy(): void {
+    this.subscriptions.unsubscribe();
+    this.resetUpdateEmailResultState();
+  }
+
+  onFormSubmit(): void {
+    if (this.form.valid) {
+      this.newUid = this.form.get('confirmEmail').value;
+      this.userService.updateEmail(this.form.get('password').value, this.newUid);
+
+      this.subscriptions.add(
+        this.getUpdateEmailResultLoading()
+          .subscribe((loading) => this.setFormControlState(loading))
+      );
+
+      this.subscriptions.add(
+        this.userService.getUpdateEmailResultSuccess()
+          .subscribe((success) => this.onSuccess(success))
+      );
+    } else {
+      this.form.markAllAsTouched();
+    }
+  }
+
+  private setFormControlState(loading: boolean): void {
+    loading ? this.form.disable() : this.form.enable();
+  }
+
+  private async onSuccess(success: boolean): Promise<void> {
```

Comment:
I'm actually not sure if we need to make the method `async`. What about just doing: 

```ts
this.authService.coreLogout().then(() => {
  // routing
})
```

Reply:
also, this logic will change with #9638, so if we won't need the `async` then as well, and can then more easily change the code without introducing a breaking change.